### DISCUSSION TITLE
Rendering custom entities with displacy: creating a custom Doc ?

### DISCUSSION THREAD

**Discussion by ThR3742** [explosion/spaCy#7239] [Help: Other Questions] [answered]
Hello and thanks for the quality of this project !

I have a text string with a set of *fixed* custom named entities (person, location, ...) as shown in the example below

```python
text = "My name is John Smith and I live in Paris"
entities = [
    ("Employee", 11, 21),  # John Smith
    ("Location", 36, 41),  # Paris
]
```

and I would like to display them using Displacy [1]. However, `displacy.render` requires a `Doc` object but I don't know how I can create such an object from my custom entities (I don't want to go through the whole nlp pipeline). Do you know what is the best way to achieve this ?

Thanks a lot for your help!

EDIT: BTW I just noticed that the new spacy 3.0 version supports my use-case (we can input ents to Doc!) but it's currently a bit hard for me to upgrade everything.

[1] https://spacy.io/usage/visualizers#ent

**Comment by polm** [comment, marked as answer]
The method on the Doc object isn't present in v2 but you can create custom spans and set the `.ents` property.

```
import spacy

nlp = spacy.blank("en")

text = "My name is John Smith and I live in Paris"
entities = [
    ("Employee", 11, 21),  # John Smith
    ("Location", 36, 41),  # Paris
]

doc = nlp(text)

ents = []
for ee in entities:
    ents.append(doc.char_span(ee[1], ee[2], ee[0]))

doc.ents = ents

for ent in doc.ents:
    print(ent, ent.label_, sep="\t")
```

**Reply by ThR3742** [reply]
Thanks a lot ! I missed the `char_span` method and didn't know we could set the ents property just like that.
